doodle-formats/src/format/deflate.rs

```diff
@@ -1204,17 +1224,26 @@ pub fn main(module: &mut FormatModule, base: &BaseModule) -> FormatRef {
         record([
             ("final", base.bit()),
             ("type", bits2.clone()),
-            (
-                "data",
-                match_variant(
+            ("data", {
+                Format::Match(
                     var("type"),
                     vec![
-                        (Pattern::U8(0), "uncompressed", uncompressed.call()),
-                        (Pattern::U8(1), "fixed_huffman", fixed_huffman.call()),
-                        (Pattern::U8(2), "dynamic_huffman", dynamic_huffman.call()),
+                        (
+                            Pattern::U8(0),
+                            fmt_variant("uncompressed", uncompressed.call()),
+                        ),
+                        (
+                            Pattern::U8(1),
+                            fmt_variant("fixed_huffman", fixed_huffman.call()),
+                        ),
+                        (
+                            Pattern::U8(2),
+                            fmt_variant("dynamic_huffman", dynamic_huffman.call()),
+                        ),
+                        // (Pattern::U8(3), Format::Fail), // Reserved - never constructed due to Fail, so no variant needed
```

Comment:
Is this a semantic change?

Reply:
There is no difference in the semantics aside from the commented-out line 1243, which would only be possible to write in the revised syntax since it doesn't need a variant and probably shouldn't get one (as Fail has no associated value to wrap in a Variant). If 1243 were uncommented it would be a semantic change, in explicitly marking type `0b11` as a failure-case rather than simply just an unexpected branch of the match-tree. Which probably wouldn't be a meaningful change aside from the error message that is returned, which are often discarded in favor of the error message returned by the final format in the NondetUnion covering the top-level alternation over each file-format.

Reply:
We might, at some point, want to distinguish strong-local errors from generic "oops this doesn't look right, try something else". Since we might, at a certain depth into a speculative parse as a given high-level Format, have high confidence that 'this is where we should be' and that specific errors should be treated as 'correct format choice, bad data' rather than 'invalidated speculative parse', which is the only real semantic class of errors we currently have a concept of.

Reply:
Oh yes I see, no need to wrap fail in a variant, that makes sense.